docs/investigation/configuration/2022-09-07-persistent-cfg-from-commands.md

```diff
@@ -0,0 +1,418 @@
+Later, we could iterate over all nested elements and check whether it should be filtered from the e.g. Communication Secret or State ConfigMap. So we'd iterate over the full object twice (once in state ConfigMap, and second time in Communications secret).
+
+## Setting configuration from BotKube commands
+
+### 1. Dedicated command for manual restart
+
+Once user changes the configuration with `@BotKube` commands, config watcher detects the changes. It doesn't quit BotKube, but just informs about config reload instead, (e.g. only the first time):
+
+> "Configuration has been updated. Apply it with the `@BotKube reload` command."
+
+The `@BotKube reload` is executed manually to apply new configuration, which would result in the app restart. This would 
+
+We don't need to change the "goodbye" and "hello" messages.
+
+Pros:
+- Very easy to implement and maintain
+- Allows to fully control the time BotKube app restarts by the user (batch multiple configuration changes)
+
+Cons:
+- UX is not great, as user needs to execute the command manually
+- Current `notifier start/stop` commands won't apply the change instantly, but the reload will be needed
+
+### 2. Restart app every time and post an updated message 
+
+Every command, even `notifier start/stop` restarts the app. But we don't post "config change detected", "goodbye" and "hello" messages.
+
+Instead, we post a notification when BotKube is online: `BotKube configuration for cluster "dev" has been reloaded 👍`.
+
+How to implement it? There would be another "state" BotKube ConfigMap, which is not monitored by Config Watcher, but still loaded during BotKube startup. Let's call it "startup-state" for now.
+
+When detecting new configuration, Config Watcher writes to the "startup-state":
+
+```yaml
+lastExitReason: "ConfigReloaded"
+```
+
+When BotKube launches again, it reads the config and posts `BotKube configuration for cluster "dev" has been reloaded 👍`. Then, it modifies the "startup-state" ConfigMap and removes the `lastExitReason` field, to not post the same message next time.
+
+Initially, it is posted for all channels as it is right now. Later, we can include the channel name in the state and post it only to the channel where the command was executed.
+
+Pros:
+- Quite easy to implement and maintain
+- BotKube reload is "hidden"
+
+Cons:
+- Current `notifier start/stop` commands will also reload BotKube. That shouldn't make big difference though, as it would be "hidden".
+
+### 3. Distinguish two types of commands and restart if necessary
+
+This is a variation of the previous approach. We could distinguish two types of commands:
+- commands that don't restart BotKube, e.g. `notifier start/stop`
+  - they save the state in runtime config and another ConfigMap, which is not monitored by Config Watcher, but still loaded during BotKube startup.
```

Comment:
> they save the state in runtime config and another ConfigMap, which is not monitored by Config Watcher

@pkosiec Can you describe how the runtime config will be used differently from the other ConfigMap?

Reply:
Currently we save the notification state into a map for each channel. E.g. for Slack:

```go
channels        map[string]channelConfigByName
```

When user executes `@BotKube notifier stop` command, we'll still write the state to the map, but also to a ConfigMap using K8sAPI. The ConfigMap will be mounted as a file as other config files, so it will be read during the BotKube startup. However it will be excluded from the files to watch by the Config Watcher: https://github.com/kubeshop/botkube/blob/main/pkg/controller/config_watcher.go#L85

Hope that answers your question 🙂

Reply:
It does - thx.